infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/CorfuLogReplicationRuntime.java

```diff
@@ -161,24 +161,24 @@ public class CorfuLogReplicationRuntime {
      */
     public CorfuLogReplicationRuntime(LogReplicationRuntimeParameters parameters,
                                       LogReplicationMetadataManager metadataManager, LogReplicationUpgradeManager upgradeManager,
-                                      LogReplicationSession session, LogReplicationContext replicationContext) {
+                                      LogReplicationSession session, LogReplicationContext replicationContext,
+                                      LogReplicationClientServerRouter router) {
         this.remoteClusterId = session.getSinkClusterId();
         this.session = session;
-        this.router = new LogReplicationClientRouter(parameters, this);
-        this.router.addClient(new LogReplicationHandler());
-        this.sourceManager = new LogReplicationSourceManager(parameters, new LogReplicationClient(router, session.getSinkClusterId()),
-                metadataManager, upgradeManager, session, replicationContext);
+        this.router = router;
+        this.sourceManager = new LogReplicationSourceManager(parameters,router, metadataManager, upgradeManager,
+                session, replicationContext);
         this.connectedNodes = new HashSet<>();
 
-        ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat("runtime-fsm-worker-"+remoteClusterId)
+        ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat("runtime-fsm-worker-"+session.hashCode())
             .build();
 
         this.communicationFSMWorkers = new ThreadPoolExecutor(1, 1, 0L,
             TimeUnit.MILLISECONDS, new LinkedBlockingQueue<>(), threadFactory);
 
         this.communicationFSMConsumer = Executors.newSingleThreadExecutor(new
                 ThreadFactoryBuilder().setNameFormat(
-                    "runtime-fsm-consumer-"+remoteClusterId).build());
+                    "runtime-fsm-consumer-"+session.hashCode()).build());
```

Comment:
Is session's hashCode include remoteClusterId? Hash represents unicode character, while remoteClusterId seems useful in debugging?

Reply:
since we now can have multiple sessions per remote cluster, it would be valuable to know which thread is working for which session. That way we can follow a particular session while debugging.

We went with hashcode because there is no easy way to name the thread with a meaningful name containing the full session information..
Currently, in sessionManager, on createSession(), we print a map of session -> hashcode, so while debugging we can look up a hashcode and follow a thread.